numpy/core/src/multiarray/item_selection.c

```diff
@@ -1074,12 +1074,34 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
                 break;
             }
         }
-        memmove(ret_data, PyArray_MultiIter_DATA(multi, mi), elsize);
+        if (out != NULL) {
+            char *args[2] = {PyArray_MultiIter_DATA(multi, mi), ret_data};
+            npy_intp transfer_strides[2] = {elsize, elsize};
+            npy_intp one = 1;
+            NPY_ARRAYMETHOD_FLAGS transfer_flags = 0;
+            NPY_cast_info cast_info = {.func = NULL};
+            PyArrayIterObject *ind_it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)out);
+            int is_aligned = IsUintAligned(ind_it->ao);
+            PyArray_GetDTypeTransferFunction(
+                        is_aligned,
+                        PyArray_DESCR(mps[0])->elsize,
+                        PyArray_DESCR(obj)->elsize,
+                        PyArray_DESCR(mps[0]),
+                        PyArray_DESCR(obj), 0, &cast_info,
+                        &transfer_flags);
+            cast_info.func(&cast_info.context, args, &one,
+                                transfer_strides, cast_info.auxdata);
+        }
```

Comment:
The boilerplate should really be outside the inner-loop, though (I guess the inputs may be unaligned, in which case we may have to use unaligned if any of them are, I don't think the output can be).  (Not sure that should be dealt with in the core, but fine.)

Also we can branch here based on references or not and remove the `PyArray_INCREF` at the end.
(If we want speed, we need to remove the `memmove` with an itemsize specialized version (like `npy_fasttake_impl`).

(I do think this whole function could probably even use the new iterator without getting too much more unwieldy, considering how unwieldy the setup is, and probably much faster)